### PR TITLE
Support topK using the DBSP operator

### PR DESCRIPTION
**Pull request by mihaibudiu**
Is this a user-visible change (yes/no): no

### CODE REVIEW

**Comment by mihaibudiu** [comment]
I will work on a commit to add more tests.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPIndexedTopKOperator.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+import javax.annotation.Nullable;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Apply a topK operation to each of the groups in an indexed collection.
+ * This always sorts the elements of each group.
+ * To sort the entire collection just group by ().
+ * The function is a DBSPSortExpression used to compare two elements.
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/circuit/operator/DBSPIndexedTopKOperator.java:19]: this is wrong, I will fix the comment in the next commit

**Comment by mihaibudiu** [comment]
Here is a sample output produced:
```
        struct Cmpstream25;
        impl CmpFunc<Tuple6<i32, F64, bool, String, Option<i32>, Option<F64>>> for Cmpstream25 {
            fn cmp(left: &Tuple6<i32, F64, bool, String, Option<i32>, Option<F64>>, right: &Tuple6<i32, F64, bool, String, Option<i32>, Option<F64>>) -> std::cmp::Ordering {
                let ord = left.1.cmp(&right.1);
                if ord != Ordering::Equal { return ord; };
                let ord = left.0.cmp(&right.0);
                if ord != Ordering::Equal { return ord; };
                let ord = left.2.cmp(&right.2);
                if ord != Ordering::Equal { return ord; };
                let ord = left.3.cmp(&right.3);
                if ord != Ordering::Equal { return ord; };
                let ord = left.4.cmp(&right.4);
                if ord != Ordering::Equal { return ord; };
                let ord = left.5.cmp(&right.5);
                if ord != Ordering::Equal { return ord; };
                return Ordering::Equal;
                }
            }
        // DBSPIndexedTopKOperator 556
        let stream25: Stream<_, OrdIndexedZSet<(), Tuple6<i32, F64, bool, String, Option<i32>, Option<F64>>, Weight>> = stream23.topk_custom_order::<Cmpstream25>(cast_to_u_i32(1i32));
```

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java [hunk 2]
```diff
@@ -1039,21 +1018,56 @@ public void visitSort(LogicalSort sort) {
...
+            if (this.options.languageOptions.ignoreOrderBy) {
+                this.assignOperator(sort, integral);
+                return;
+            }
+            // Otherwise we have to sort again in a vector!
+            // Fall through, continuing from the integral.
+            this.circuit.addOperator(integral);
+            index = integral;
+        }
+        // Global sort.  Implemented by aggregate in a single Vec<> which is then sorted.
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/frontend/CalciteToDBSPCompiler.java:1040]: this code is unchanged

**Comment by ryzhyk** [comment]
> Here is a sample output produced:
> 
> ```
>         struct Cmpstream25;
>         impl CmpFunc<Tuple6<i32, F64, bool, String, Option<i32>, Option<F64>>> for Cmpstream25 {
>             fn cmp(left: &Tuple6<i32, F64, bool, String, Option<i32>, Option<F64>>, right: &Tuple6<i32, F64, bool, String, Option<i32>, Option<F64>>) -> std::cmp::Ordering {
>                 let ord = left.1.cmp(&right.1);
>                 if ord != Ordering::Equal { return ord; };
>                 let ord = left.0.cmp(&right.0);
>                 if ord != Ordering::Equal { return ord; };
>                 let ord = left.2.cmp(&right.2);
>                 if ord != Ordering::Equal { return ord; };
>                 let ord = left.3.cmp(&right.3);
>                 if ord != Ordering::Equal { return ord; };
>                 let ord = left.4.cmp(&right.4);
>                 if ord != Ordering::Equal { return ord; };
>                 let ord = left.5.cmp(&right.5);
>                 if ord != Ordering::Equal { return ord; };
>                 return Ordering::Equal;
>             }
>         }
>         // DBSPIndexedTopKOperator 556
>         let stream25: Stream<_, OrdIndexedZSet<(), Tuple6<i32, F64, bool, String, Option<i32>, Option<F64>>, Weight>> = stream23.topk_custom_order::<Cmpstream25>(cast_to_u_i32(1i32));
> ```

Looks good, assuming it's all `order by .. ASC` in SQL.

**Review by ryzhyk — APPROVED** [review]

**Comment by mihaibudiu** [comment]
Right, have to handle descending. Easy.

**Comment by ryzhyk** [comment]
BTW, I think this is a user-visible change, which requires docs + a changelog entry.

**Comment by mihaibudiu** [comment]
which part is visible? LIMIT was there in the grammar.

**Comment by ryzhyk** [comment]
> which part is visible? LIMIT was there in the grammar.

Clearly, we're adding a new user-visible capability.

**Comment by mihaibudiu** [comment]
No, we are fixing a bug

**Comment by ryzhyk** [comment]
> No, we are fixing a bug

So you're saying we're implementing a behavior that documentation already promised before.  If the documentation is accurate, then we're good.  A changelog entry documenting the bug fix is still in order.

File: sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/util/TableValue.java [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+import static org.dbsp.sqlCompiler.ir.type.DBSPTypeCode.USER;
```
Comment (mihaibudiu) [inline sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/util/TableValue.java:22]: This code was moved here from DBSPExecutor

**Pull request merged**